rdmd.d

```diff
@@ -916,3 +900,41 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function. These functions allow the the filenane to be given once
+so the log statement always matches the operation. Using it also guarantees
+you won't forget to include a `yap` alongside any file operation you want
+to be logged.
+*/
+static struct Yap
+{
+    static auto opDispatch(string op, T...)(T args)
+    {
+        static if (op == "exists" ||
+                   op == "timeLastModified" ||
+                   op == "isFile")
+            yap("stat ", args[0]);
+        else static if (op == "rename")
+            yap("mv ", args[0], " ", args[1]);
+        else static if (op == "remove" ||
+                        op == "rmdirRecurse" ||
+                        op == "dirEntries")
+            yap(op, " ", args[0]);
+        else static assert(0, "Yap.opDispath has not implemented " ~ op);
+
+        mixin("return std.file." ~ op ~ "(args);");
+    }
+    static bool existsAsFile(const(char)[] name)
+    {
+        yap("stat ", name);
+        return std.file.exists(name) && std.file.isFile(name);
+    }
+    static void mkdirRecurseIfLive(const(char)[] name)
```

Comment:
Oof, this is not good. There is a hole in the design: first `mkdirRecurseIfLive`, and then it looks like `removeIfLive` is now needed.

The "if live" thing needs to either become a parameter (which either switches between "log only / log and do" or "log then run only if `dryRun` is false") or possibly inferred from something (e.g. maybe we can assume that if a function returns `void` then it mutates external state and should not be called when `dryRun` is true, but that fails for e.g. `getTimes`).

Reply:
Maybe something like

```d
struct FilesystemImpl(bool always)
{
    auto fun(...)
	{
		yap(...);
		if (always || !dryRun)
			return std.file.fun(...);
	}
}

alias AlwaysDo = FilesystemImpl!true;
alias MaybeDo = FilesystemImpl!false;

bool fileExists = AlwaysDo.exists(fileName);

MaybeDo.mkdirRecurse(dirPath);
```

Probably needs better names...

Reply:
Heh, I guess that's one way to do it... but, looking at the identifiers and doing string operations with them... well, I guess that might very well be the least evil option.